test/e2e/scheduling/nvidia-gpus.go

```diff
@@ -103,13 +103,13 @@ func areGPUsAvailableOnAllSchedulableNodes(f *framework.Framework) bool {
 	return true
 }
 
-func getGPUsAvailable(f *framework.Framework) int64 {
+func getGPUNodesAvailable(f *framework.Framework) int64 {
```

Comment:
I can see that `makeCudaAdditionDevicePluginTestPod` creates a pod with 2 containers each requesting one GPU and that may not work well if nodes have only one GPU, but doesn't the number of containers created should match the number GPUs (as opposed to the number of nodes with GPU)?

Reply:
Yes, the number of containers (where each requests 1 GPU) should match the number of GPUs. However, this function is being used to create pods, which is why we only want the number of nodes, as opposed to GPUs. https://github.com/kubernetes/kubernetes/blob/21c83d397ce55f7e17a461ea64ced6f2b27c93e5/test/e2e/scheduling/nvidia-gpus.go#L168

Reply:
My point is that the new logic does not look perfect either. Do we know how many GPUs each node has, or can we control it? If each node has 10 GPUs, a few pods may be scheduled on the same node and the test may pass, while there are other nodes that are unschedulable while they should have been schedulable.
A better approach is to find out the number of GPUs per node and then create pods accordingly. For example, if each node has 2 GPUs, we create the same number of pods as the nodes. If each node has only 6 GPUs, we may want to add node affinity to the pods to ensure that they are scheduled on all the available nodes. Alternatively, we can add inter-pod anti-affinity to the pods to make sure that they are not scheduled on the same node.

Reply:
Thanks for your comment.

I've updated the test to divide the number of GPUs by containers per pod, which will remain at 2, as discussed #77151.

Regarding number of GPUs on the nodes, we can control how many GPUs are on the node in the test-infra config, such as here:
https://github.com/kubernetes/test-infra/blob/bfaec0219165084674cc393ac5e55741a8a9aff1/config/jobs/kubernetes/sig-gcp/gpu/sig-gcp-gpu-gke.yaml#L21